### PR TITLE
Article cdn fastly cloudfront

### PR DESCRIPTION
**Pull request by mezis**
First post!

cc @jphastings, our beloved Editor-in-Chief. Feel free to rewrite any part of this.

To read the article locally:

- Add `future: true` to `_config.yml`
- Run `bundle exec jekyll serve`

### CODE REVIEW

File: _posts/2016-07-13-split-testing-cdns.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+  Split testing is a cornerstone of how we improve our products. While we
+  usually run such tests for user-visible interface changes, this is an example
+  of running a (successful!) multivariate test between CDNs.
+
+---
+
+TL,DR:
+
+- Fastly is generally more performant than AWS Cloudfront — for our userbase.
+- A/B testing CDNs is easy (with out toolset).
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: with out => without/with our?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+We split tested Cloudfront vs. Fastly across 1,000,000 users, in 12
+countries, on 3 continents, over a significant time period. Using Fastly
+improved page load times by up to 50% and site conversion by up to a whopping
+3.9% (with _p_ < 0.01). Read on for the caveats!
+
+---
+
+## Context
+
+We currently use a CDN[^cdn]
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: This doesn't seem to render right? Or is it just the markdown preview on github acting funny?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+
+We'll come back to the details of this chart a bit later, but it should pretty
+clearly show that for Cloudfront (or any CDN, really), actual performance from
+actual users varies quite a bit by geographical market - sometimes dramatically
+so, as with Australia.
+
+
+Our current HTTP infrastructure has 3 tiers:
+
+- Client requests hit either a CDN (assets) or our reverse proxy (NGinx).
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: *Nginx

File: _posts/2016-07-13-split-testing-cdns.md [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+- The CDN loads from the proxy on misses.
+- The proxy is backed by a Heroku-hosted application.
+
+Both the proxy and the Heroku app are hosted by AWS, in the `eu-west-1` (Dublin)
+datacenter.
+
+## Testable instructure variants
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: *infrastructure

File: _posts/2016-07-13-split-testing-cdns.md [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+Because Cloudfront is our existing setup, We allowed for a 1-week warmup period:
+we wouldn't want to compare another CDN cache until we were confident that most
+POPs had had a chance to cache the asset in question.
+
+We then let the system harvest data for a while, and ran analysis queries in [NR
+Insights][nrinsights].
+
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--03.png)
+{: .dg-figure}
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: Same as the footnote above and the one below

File: _posts/2016-07-13-split-testing-cdns.md [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+hypothesis: in a given country, for a given device type, Cloudfront and Fastly
+perform the same.
+
+The metric we'll look at is the duration between the "time to first byte" (the
+point at which, when loading a page, a browser starts receiving the HTTP
+response) and the "page loaded" event (when
+all synchronous assets have been laoded and the page has been rendered).
+
+This is the part of the page load lifecycle where differences in asset load
+timings may have an impact. Given all other things are unchanges, it's the
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: *unchanged

File: _posts/2016-07-13-split-testing-cdns.md [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+title:  "Running A/B tests on our hosting infrastructure"
+date:   2016-07-13 00:00:00
+exerpt: >
+  Split testing is a cornerstone of how we improve our products. While we
+  usually run such tests for user-visible interface changes, this is an example
+  of running a (successful!) multivariate test between CDNs.
+
+---
+
+TL,DR:
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Having a TL,DR; is great, but it'd be great to frame the article with a short paragraph first. Call me buzz-feed generation, but it'd be good to build some excitement about the article here & give folks a reason to read it, instead of a reason to not read it 😛

File: _posts/2016-07-13-split-testing-cdns.md (hunk 1 above)
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/out/our/`?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+sources that we know correlate to conversion. For user interfaces, this can be
+focus groups early on and user research down the line; for infrastructure,
+performance measurements and synthetic checks won't be displaced by A/B testing
+anytime soon!
+
+## Tools used
+
+- [Amazon Cloudfront][cloudfront] and [Fastly][fastly]: the two CDNs we've
+  pitted against eachother.
+- New Relic [Synthetics][nrsynth]: a Pindgom clone that lest up repeatedly call
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: lest up => let's us?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+  and HTTP endpoint and get statistics on response time.
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: and => an

File: _posts/2016-07-13-split-testing-cdns.md [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+
+Now that we've instructed our backend to serve different asset URLs for
+different test variants, we need the ability to report on performance for each
+variant.
+
+We use [NR Browser][nrbrowser], which injects a bit of Javascript in each HTML
+page and reports on performance — somewhat like a dumbed-down
+[Webpagetest](http://www.webpagetest.org/), but on every single page view.
+
+The JS library lets us "tag" each pageview with a metadata that we can later
```
Comment (tommyp) [inline _posts/2016-07-13-split-testing-cdns.md]: `a metadata`?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+---
+
+TL,DR:
+
+- Fastly is generally more performant than AWS Cloudfront — for our userbase.
+- A/B testing CDNs is easy (with out toolset).
+- Synthetic performance checks (e.g. with Pingdom) do not reflect end user
+  performance experience.
+
+We split tested Cloudfront vs. Fastly across 1,000,000 users, in 12
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: This sentence is an impressive statement to make; I'm not sure if Jekyll give us easy-to-use pull-quotes but I'd love to make this more prominent; it might be a bit too cheesy, but we could embolden the numbers out so they stand out. What do you think of something like this?

> We split our production traffic between Cloudfront and Fastly's CDN networks, looking for significant (_p_ < 0.01) differences in page load times and conversion across **1,000,000 users** in **12 countries** on **3 continents**. We found Fastly a massive **50% faster** for page load times, which translated into a whopping **3.9% increase in conversion**.
>
> Naturally there are some important caveats to these statistic, and we think the processes we've used to run this experiment are interesting enough to blog about, so read on for the detail…

File: _posts/2016-07-13-split-testing-cdns.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+countries, on 3 continents, over a significant time period. Using Fastly
+improved page load times by up to 50% and site conversion by up to a whopping
+3.9% (with _p_ < 0.01). Read on for the caveats!
```
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: Those are some pretty serious caveats though 😸 It is quite a bit misleading though, I think it sets up very high expectations compared to the outcome in the end. Which is basically there's a marginal increase in performance outside of AU and HK and the conversion confidence intervals are quite all over the place which from my entirely vague understanding of statisticsI wouldn't count as statistically significant increases in conversion apart from DE and slightly for FR which is also a bit suspicious because they are not among the top performance increases. I'm just pointing it out because the abstract seems to link conversion to performance which is not what you're claiming throughout the article. You may have done this on purpose but that's my 2 cens

File: _posts/2016-07-13-split-testing-cdns.md (hunk 2 above)
Comment (tommyp) [inline _posts/2016-07-13-split-testing-cdns.md]: I think it's the latter.

File: _posts/2016-07-13-split-testing-cdns.md (hunk 5 above)
Comment (mezis) [inline _posts/2016-07-13-split-testing-cdns.md]: Github Pages uses [Kramdown](https://help.github.com/articles/updating-your-markdown-processor-to-kramdown/) — the renderer in PRs uses [GFM](https://help.github.com/articles/about-writing-and-formatting-on-github/).

Welcome to the world of concurrent Markdown implementations.

File: _posts/2016-07-13-split-testing-cdns.md (hunk 11 above)
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: The above does accentuate the page load -> conversion link which @sdorunga1 has questions about, it may not be an appropriate link to make!

File: _posts/2016-07-13-split-testing-cdns.md [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+- The proxy is backed by a Heroku-hosted application.
+
+Both the proxy and the Heroku app are hosted by AWS, in the `eu-west-1` (Dublin)
+datacenter.
+
+## Testable instructure variants
+
+We set up 4 variants of our infrastrucutre:
+`cdn0` is the existing, Cloudfront-based setup - the control.
+`cdn1` to `cdn3` to replaces Cloudfront with various Fastly configurations:
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/to replaces/replaced/`

File: _posts/2016-07-13-split-testing-cdns.md [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+
+- `cdn1` uses Fastly's "shield" feature, whereby local POPs[^pop] are backed by a
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Link to product page for the shield feature?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+  performance experience.
+
+We split tested Cloudfront vs. Fastly across 1,000,000 users, in 12
+countries, on 3 continents, over a significant time period. Using Fastly
+improved page load times by up to 50% and site conversion by up to a whopping
+3.9% (with _p_ < 0.01). Read on for the caveats!
+
+---
+
+## Context
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Something more engaging?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--02.png)
+{: .dg-figure}
+
+
+
+### Synthetic performance comparison
+
+To get a sense of what performance differences we could observe between
+countries, we set up a synthetic check per test variant in NR Synthetics.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `[NR Synthetics](#tools-used)`?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+
+Each of these checks is configured to request a given asset (our main CSS file)
+every minute, from a number of different locations across the globe; NR Insights
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `[NR Insights](#tools-used)`?

File: _posts/2016-07-13-split-testing-cdns.md (hunk 7 above)
Comment (mezis) [inline _posts/2016-07-13-split-testing-cdns.md]: The exerpt (in the Yaml front matter) will be visible both in the list of posts on on the post page — is that what you mean?

I can make that intro more intro-y if you want though.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+`cdn1` to `cdn3` to replaces Cloudfront with various Fastly configurations:
+
+- `cdn1` uses Fastly's "shield" feature, whereby local POPs[^pop] are backed by a
+  second-tier cache (also managed by Fastly). This cache is located in London —
+  as close as possible to our main datacenter.
+- `cdn2` uses our current reverse proxy (which has some in-memory caching) as
+  the CDN's upstream. This most closely mimics our Cloudfront setup.
+- `cdn3` is backed directly by the Heroku application.
+
+Each of those is given a different DNS entry for each of our TLDs,
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: You've swapped tenses here - "we **set up** 4 variants" vs "Each of those **is** given" - I think the past tense was working for you, could we stick with that?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+{: .dg-figure}
+
+
+
+### Synthetic performance comparison
+
+To get a sense of what performance differences we could observe between
+countries, we set up a synthetic check per test variant in NR Synthetics.
+
+Each of these checks is configured to request a given asset (our main CSS file)
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: If we're going with past tense this `is` should be a `was`.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+every minute, from a number of different locations across the globe; NR Insights
+lets us access detailed reports of performance for each individual request.
+
+Because Cloudfront is our existing setup, We allowed for a 1-week warmup period:
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/We/we/`

File: _posts/2016-07-13-split-testing-cdns.md [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+POPs had had a chance to cache the asset in question.
+
+We then let the system harvest data for a while, and ran analysis queries in [NR
+Insights][nrinsights].
+
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--03.png)
+{: .dg-figure}
+
+At first glance, this looks like Cloudfront (`cdn0`) is massively winning over
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: "this" is slightly unspecified, can we go with a clearer subject, eg. "At first glance this graph seems to show that Cloudfront (`cdn0`) …"?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+Fastly (`cdn[1-3]`). The trick here is that Cloudfront is part of Amazon's
+infrastructure, so it's unfairly advantaged whenever the pinging server is also
+in AWS — most of the cases here.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: I'm not quite sure what "most of the cases here" is trying to say - can we expand it a little bit?

File: _posts/2016-07-13-split-testing-cdns.md (hunk 12 above)
Comment (mezis) [inline _posts/2016-07-13-split-testing-cdns.md]: Actually, the performance improvement is pretty massive locally, and very good globally.
It's on _conversion_ that we don't see the expected improvements.

I'll try to rephrase a bit, see if I can convince you.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+```html
+<script type="text/javascript">
+  if (typeof newrelic !== "undefined") {
+    newrelic.setCustomAttribute("cdnServer", "<%= $cdn %>");
+  }
+</script>
+```
+
+From thereon, it's just a matter of waiting enough time for enough data to be
+harvested and reaching statistical significance.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: "for enough data to be harvested_to_ reaching statistical significance."?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+> latency. It's best if the CDN has a POP at the user's ISP, and ISP networks are
+> typically per-country with a few exceptions (dense areas like Europe).
+> Similarly, DSL/fibre ISPs may not share a network with mobile/GSM ISPs.
+>
+> As we cannot split the data on the "type" of network (landline vs. mobile),
+> we're splitting on device types as a proxy. This is determined by interpreting
+> the user-agent header for each HTTP request.
+{: .dg-sidebar}
+
+For this second test, we'll run a proper A/B test trying to falsify the
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `we'll run` is future tense - we should probably stick to the past as above?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+</script>
+```
+
+From thereon, it's just a matter of waiting enough time for enough data to be
+harvested and reaching statistical significance.
+
+
+### Performance for actual users
+
+> ##### Why split the analysis by country _and_ platform?
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Might be worth moving this section below the next paragraph, as on mobile these sections will appear in-line (#12) and the the main paragraph introduces the decision to do the thing this one is explaining :)

File: _posts/2016-07-13-split-testing-cdns.md [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+perform the same.
+
+The metric we'll look at is the duration between the "time to first byte" (the
+point at which, when loading a page, a browser starts receiving the HTTP
+response) and the "page loaded" event (when
+all synchronous assets have been laoded and the page has been rendered).
+
+This is the part of the page load lifecycle where differences in asset load
+timings may have an impact. Given all other things are unchanges, it's the
+_only_ factor of change, which is what we're after in an A/B test.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: These seems a little tautological: if everything else is unchanged then this is the only thing that changed. I'm not quite sure what point you were trying to make beyond that experiments should be limited to only one dynamic variable (an excellent point to make!) so I'm not sure how I'd rephrase this - any ideas?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+  , stddev(domProcessingDuration + pageRenderingDuration) AS 'Stddev'
+  , percentile(domProcessingDuration + pageRenderingDuration, 50) AS 'Time'
+FROM PageView
+WHERE deviceType = 'Desktop'
+FACET countryCode
+SINCE '2016-07-01'
+LIMIT 20
+```
+
+That's not extremely convenient, and we'll need to post-process anyhow (to run
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: "The HTML output isn't very convenient but we needed to post-process (to run …), so …" makes the subject of this sentence much clearer ("That" probably isn't the right determiner here, you'd want to use "this", but an explicit subject is nice as the reader has just been deep in code 😊)

File: _posts/2016-07-13-split-testing-cdns.md [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+the [T-Test](https://en.wikipedia.org/wiki/Student%27s_t-test) math), so half an
+hour or copying-and-pasting into Excel follows.
+
+Intermission: for proper A/B testing, have a read of [Evan Miller's Awesome
+A/B Testing Tools](http://www.evanmiller.org/ab-testing/). One of my favourite
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: :+1:

File: _posts/2016-07-13-split-testing-cdns.md [hunk 29]
```diff
@@ -0,0 +1,352 @@
...
+bookmarks.
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--04.png)
+{: .dg-figure}
+
+This graphs the ratio of load+render times for Fastly (`cdn1` variant, the best
+one) versus Cloudfront (`cdn0`), split by the users' countries - anything on the
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: "(`cdn0` variant, our pre-existing set up)" for balance?

File: _posts/2016-07-13-split-testing-cdns.md [hunk 30]
```diff
@@ -0,0 +1,352 @@
...
+right, Fastly wins.
+
+The error bars are for _p < 0.05_, not good enough for Science or Nature but
+generally accepted as good enough for multivariate tests in tech companies.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: If there's a reasonable citation you can offer here it'd be really interesting, though I don't think it's needed for the article :)

File: _posts/2016-07-13-split-testing-cdns.md [hunk 31]
```diff
@@ -0,0 +1,352 @@
...
+
+Unlike for our earlier synthetic tests, the results are quite clear cut:
+
+- Fastly is significantly faster almost everywhere.
+- Notable exceptions exist: the UAE (Fastly doesn't ahve a POP there yet); Spain
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/ahve/have/`

File: _posts/2016-07-13-split-testing-cdns.md [hunk 32]
```diff
@@ -0,0 +1,352 @@
...
+  on desktops (no obvious explanation), Ireland and Singapore on mobiles (Fastly
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: This was a semicolon delimited list a moment ago - can we `s/,/;/`? 😊

File: _posts/2016-07-13-split-testing-cdns.md [hunk 33]
```diff
@@ -0,0 +1,352 @@
...
+  probably has a weaker presence there than Cloudfront — likely as the capitals
+  there are Amazon datacenter cities).
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: I'd be tempted to break these exceptions out into their own sub-bullet points so you can address them each more clearly (rather than in parentheses) - this last one is particularly interesting!

File: _posts/2016-07-13-split-testing-cdns.md [hunk 34]
```diff
@@ -0,0 +1,352 @@
...
+
+At this point we could opt to jsut switch over to Fastly and bite the bullet in
+places where it's slightly weaker; but what ultimately matters is not how fast
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Italics for emphasis? I'm in two minds…

> but what ultimately matters is not _how fast_ our site and apps are, but _how many users we satisfy_.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 35]
```diff
@@ -0,0 +1,352 @@
...
+our site and apps are, but how many users we satisfy.
+
+
+### Impact on conversion
+
+For this last test, we're using the same A/B test setup as above, but this time
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Tense here again; I don't mind swapping to present for the article as you seem to favour it in the latter parts, but a consistent one would be good :)

File: _posts/2016-07-13-split-testing-cdns.md (hunk 25 above)
Comment (mezis) [inline _posts/2016-07-13-split-testing-cdns.md]: Looks terrible on desktop - but point taken. Will move it.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 36]
```diff
@@ -0,0 +1,352 @@
...
+we're measuring conversion across CDN variants, countries, and platforms.
+
+We use [Segment][segment] to relay and store our analytics. The gist is that
+form anywhere in our backend, we can say:
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: > The gist is that form anywhere in our backend, we can say:
This sounds a little strange, maybe something like the following?
> It allows us to track events, including custom attributes, from anywhere in our back end application:

Also, if you aim to keep the original `s/form/from/`

File: _posts/2016-07-13-split-testing-cdns.md (hunk 25 above)
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Hmmm. May be worth leaving until after we know how we're going to tackle #12 then!

File: _posts/2016-07-13-split-testing-cdns.md [hunk 37]
```diff
@@ -0,0 +1,352 @@
...
+
+```ruby
+$segment.track 'Session Started', cdn: $cdn
+# (...)
+$segment.track 'Completed Order', cdn: $cdn
+```
+
+and end up with a record in Redshift, tagged with the user who made the request,
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: I'm not such a fan of code blocks being part of a sentence like this; it doesn't flow so well for people unfamiliar with the code you're presenting. You could restart the text with something like "This inserts a record in Redshift…"

File: _posts/2016-07-13-split-testing-cdns.md [hunk 38]
```diff
@@ -0,0 +1,352 @@
...
+the browser, the country, the CDN variant used, and a whole lot of other bits of
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Parenthesise the last point for clarity?

> the country and the CDN variant used (as well as a whole lot of other bits of metadata we won't use here.)

File: _posts/2016-07-13-split-testing-cdns.md [hunk 39]
```diff
@@ -0,0 +1,352 @@
...
+metadata we won't use here.
+
+Again, once that's set up, it's just a matter of waiting until there's enough
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: it's is present tense, if you're changing to past :)

File: _posts/2016-07-13-split-testing-cdns.md [hunk 40]
```diff
@@ -0,0 +1,352 @@
...
+data to run a significant [χ²
+test](https://en.wikipedia.org/wiki/Chi-squared_test).
+
+Overall, the test results are reassuringly consistent with the changes in
+performance: depending on the Fastly setup (`cdn[1-3]`), we observe either no
+conversion change (+0.0% ±0.8%, _p_ = 0.5), or a slight improvement (+0.9%
+±0.9%, _p_ = 0.02).
+
+But once split by country and platform results are surprising.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: We need a break in there to aid reading, something like this? 

> But once split by country and platform the results are surprising.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 41]
```diff
@@ -0,0 +1,352 @@
...
+
+
+![Infrastructure variants](/images/2016-07-13-split-testing-cdns--05.png)
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Title of graph has `s/when/with/` I believe

File: _posts/2016-07-13-split-testing-cdns.md [hunk 42]
```diff
@@ -0,0 +1,352 @@
...
+{: .dg-sidebar}
+
+Error bars here are still for _p_ < 0.05, so any result where the error bar
+crosses zero can be discarded. This leaves us with:
+
+- Germany and France see a significant improvement in both performance and
+  conversion.
+- Most other countries have conversion and performance trending in the same
+  direction.
+- But... while Australian mobiles see a significant improvement in performance
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `while … but` sadly doesn't scan; we can replace `s/ but/,/` or `s/while //` (we can't use both forms at once).

File: _posts/2016-07-13-split-testing-cdns.md [hunk 43]
```diff
@@ -0,0 +1,352 @@
...
+  but conversion is significantly worse!
+
+We don't have an explanation for this last point. It's possible that major
+Aussie ISPs do something bad to our HTTP requests that cuase some to fail, but
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/cuase/cause/`

File: _posts/2016-07-13-split-testing-cdns.md [hunk 44]
```diff
@@ -0,0 +1,352 @@
...
+we haven't been able to reproduce error cases or received error reports form our
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Generally an 'or' split like this implies the use of the same verb before the latter part, which implies "we haven't been able to received error reports", making this sound a little odd. You could add the pronoun again to make it clearer: "able to reproduce error cases nor have we received error".

File: _posts/2016-07-13-split-testing-cdns.md [hunk 45]
```diff
@@ -0,0 +1,352 @@
...
+local teams or local customers.
+
+This leave us with our heads scratching, and another example that AB tests
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/leave/leaves/`

File: _posts/2016-07-13-split-testing-cdns.md [hunk 46]
```diff
@@ -0,0 +1,352 @@
...
+aren't perfect: they can give strong results... but they can also give very
+inconsistent results if there's a hidden variable.
+
+Our take on this is to, as we did above, couple A/B testing with other
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: I'd stick the subclause after the content of the sentence it's hard for the reader to reference what we did above when they don't know the context of the sentence is yet:

> Our take on this was to couple A/B testing with other sources that we know correlate to conversion, as we did above.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 47]
```diff
@@ -0,0 +1,352 @@
...
+sources that we know correlate to conversion. For user interfaces, this can be
+focus groups early on and user research down the line; for infrastructure,
+performance measurements and synthetic checks won't be displaced by A/B testing
+anytime soon!
+
+## Tools used
+
+- [Amazon Cloudfront][cloudfront] and [Fastly][fastly]: the two CDNs we've
+  pitted against eachother.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: `s/eachother/each other/` for the pedantic amongst… me 😛

File: _posts/2016-07-13-split-testing-cdns.md (hunk 46 above)
Comment (mezis) [inline _posts/2016-07-13-split-testing-cdns.md]: I'm rewriting this section in part so - maybe have another read after?

**Comment by jphastings** [comment]
Really interesting! Thanks Julien, it's a great article. I've made a couple of notes here and there with one general point around going with a single tense for the article (past or present I'd recommend - though bonus points for something wacky 😛)

**Comment by mezis** [comment]
@jphastings, I think I've addressed most of your comments.
I've also revamped the intro and conclusion, and added a couple graphs to address @sdorunga1's valid concerns.

File: _posts/2016-07-13-split-testing-cdns.md (hunk 12 above)
Comment (sdorunga1) [inline _posts/2016-07-13-split-testing-cdns.md]: Yep, that's basically what I meant. The way the text was worded it did make it sound like a very tight and quite high relationship existed between performance and conversion (which is what is currently believed to generally be true) which would form a very different expectation in the mind of a reader.

I'm very happy with the new wording as it's making it a lot clearer what the outcome actually was, why it's significant/useful and how you came up with those conclusions which is really valuable.

**Comment by sdorunga1** [comment]
Very good articl! I think showing exactly how we came up with the numbers and the very detailed explanations of what you chose to ignore/take into account were very clear and useful. Especially compared with just having a load of data presented as graphs and a vague conclusion which is what most articles do :)

Are we still running the AB test btw?

**Comment by mezis** [comment]
@sdorunga1: Thanks! Yeah, we're still running the test. I'll probably leave it up until we're ready to go properly live with Fastly (sorting out proper SSL certs for now), because, why not get more data and update the graphs.

@jphastings: as I'm on hols next week, can I leave this PR with you? You're welcome to change anything. Seems to me it's almost good to merge.

File: _posts/2016-07-13-split-testing-cdns.md [hunk 48]
```diff
@@ -0,0 +1,408 @@
...
+We'll come back to the details of this chart a bit later, but it should pretty
+clearly show that for Cloudfront (or any CDN, really), actual performance from
+actual users varies quite a bit by geographical market - sometimes dramatically
+so, as with Australia.
+
+
+Our current HTTP infrastructure has 3 tiers:
+
+- Client requests hit either a CDN (assets) or our reverse proxy
+  ([Nginx][nginx]).
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: No longer links anywhere :(

File: _posts/2016-07-13-split-testing-cdns.md [hunk 49]
```diff
@@ -0,0 +1,408 @@
...
+- The CDN loads from the proxy on misses.
+- The proxy is backed by a [Heroku][heroku]-hosted application.
```
Comment (jphastings) [inline _posts/2016-07-13-split-testing-cdns.md]: Nor this one :()

**Comment by jphastings** [comment]
@mezis I left it until you were back in the office (welcome back! 🏖) but I think we're almost ready to publish - some mini issues I've commented on!

As an aside, all our external links are loading in-window (rather than a new tab/window), can we change this and make it part of Jekyll's config?

**Comment by mezis** [comment]
> As an aside, all our external links are loading in-window (rather than a new tab/window), can we change this and make it part of Jekyll's config?

I don't think we can, no. Unfortunately. This said blogs that open all links in a new tab are a slightly cringe-worthy experience (to me!)
